src/iccpd/src/iccp_csm.c

```diff
@@ -0,0 +1,802 @@
+    csm->iccp_info.sender_capability_flag = 0x0;
+    csm->iccp_info.sender_rg_connect_flag = 0x0;
+    app_csm_init(csm, all);
+
+    memset(&csm->msg_log, 0, sizeof(struct MsgLog));
+}
+
+/* Connection State Machine instance tear down */
+void iccp_csm_finalize(struct CSM* csm)
+{
+    struct If_info * cif = NULL;
+    struct System* sys = NULL;
+
+    if (csm == NULL)
+        return;
+
+    if ((sys = system_get_instance()) == NULL)
+        return;
+
+    /*If warm reboot, don't change port block and peer link MAC learning*/
+    if (sys->warmboot_exit != WARM_REBOOT)
+    {
+        /* Clean all port block*/
+        peerlink_port_isolate_cleanup(csm);
+
+        /*Enable peer link port MAC learning*/
+        set_peerlink_mlag_port_learn(csm->peer_link_if, 1);
+    }
+
+    /* Disconnect from peer */
+    scheduler_session_disconnect_handler(csm);
+
+    /* Release all Connection State Machine instance */
+    app_csm_finalize(csm);
+
+    LIST_FOREACH(cif, &(csm->if_bind_list), csm_next)
+    {
+        LIST_REMOVE(cif, csm_next);
+    }
+
+    /* Release iccp_csm */
+    pthread_mutex_destroy(&(csm->conn_mutex));
+    iccp_csm_msg_list_finalize(csm);
+    LIST_REMOVE(csm, next);
+    free(csm);
+}
+
+/* Message list of Connection State Machine instance tear down */
+void iccp_csm_msg_list_finalize(struct CSM* csm)
+{
+    struct Msg* msg = NULL;
+
+    if (csm == NULL)
+        return;
+
+    while (!TAILQ_EMPTY(&(csm->msg_list)))
+    {
+        msg = TAILQ_FIRST(&(csm->msg_list));
+        TAILQ_REMOVE(&(csm->msg_list), msg, tail);
+        free(msg);
+    }
+}
+
+/* Send message to peer */
+int iccp_csm_send(struct CSM* csm, char* buf, int msg_len)
+{
+    LDPHdr* ldp_hdr = (LDPHdr*)buf;
+    ICCParameter* param = NULL;
+
+    if (csm == NULL || buf == NULL || csm->sock_fd <= 0 || msg_len <= 0)
+        return -1;
+
+    if (ntohs(ldp_hdr->msg_type) == MSG_T_CAPABILITY)
+        param = (struct ICCParameter*)&buf[sizeof(LDPHdr)];
+    else
+        param = (struct ICCParameter*)&buf[sizeof(ICCHdr)];
+
+    /*ICCPD_LOG_DEBUG(__FUNCTION__, "Send(%d): len=[%d] msg_type=[%s (0x%X, 0x%X)]", csm->sock_fd, msg_len, get_tlv_type_string(param->type), ldp_hdr->msg_type, param->type);*/
+    csm->msg_log.msg[csm->msg_log.end_index].msg_id = ntohl(ldp_hdr->msg_id);
+    csm->msg_log.msg[csm->msg_log.end_index].type = ntohs(ldp_hdr->msg_type);
+    csm->msg_log.msg[csm->msg_log.end_index].tlv = ntohs(param->type);
+    ++csm->msg_log.end_index;
+    if (csm->msg_log.end_index >= 128)
+        csm->msg_log.end_index = 0;
+
+    return write(csm->sock_fd, buf, msg_len);
+}
+
+/* Connection State Machine Transition */
+void iccp_csm_transit(struct CSM* csm)
+{
+    int len = -1;
+    struct Msg* msg = NULL;
+    ICCP_CONNECTION_STATE_E prev_state;
+
+    if (!csm)
+        return;
+
+    prev_state = csm->current_state;
+
+    /* No connection, but have state change? reset it...*/
+    if (csm->current_state != ICCP_NONEXISTENT && csm->sock_fd <= 0)
```

Comment:
i am trying to understand this case. what is this scenario where is a iccp state change and there is no connection. please help me understand

Reply:
This is a fault-tolerance. If csm->sock_fd socket is closed for some reason, change the state back to ICCP_NONEXISTENT.

Reply:
ok thanks